tsinfer/inference.py

```diff
@@ -890,23 +978,51 @@ def __init__(
             [inference_site_position, [sample_data.sequence_length]]
         )
         self.position_map[0] = 0
+        self.recombination = np.zeros(self.num_sites)  # TODO: reduce len by 1
+        self.mismatch = np.zeros(self.num_sites)
+
+        if recombination is not None or mismatch is not None:
+            if recombination is None or mismatch is None:
+                raise ValueError(
+                    "Directly setting probabilities requires specifying "
+                    "both 'recombination' and 'mismatch'"
+                )
+            if recombination_rate is not None or mismatch_ratio is not None:
+                raise ValueError(
+                    "Cannot simultaneously specify recombination & recombination_rate, "
+                    "or mismatch and mismatch_ratio"
+                )
+        else:
+            # Must set recombination and mismatch arrays
+            if recombination_rate is None and mismatch_ratio is not None:
+                raise ValueError("Cannot use mismatch without setting recombination")
+            if (
+                recombination_rate is None and mismatch_ratio is None
+            ) or num_intervals == 0:
+                # Special case: revert to tsinfer 0.1 behaviour with no mismatch allowed
+                default_recombination_prob = 1e-2
+                default_mismatch_prob = 1e-20  # Substantially < the value above
+                recombination = np.full(num_intervals, default_recombination_prob)
+                mismatch = np.full(self.num_sites, default_mismatch_prob)
+            else:
+                recombination = recombination_rate_to_probs(
+                    recombination_rate, inference_site_position
+                )
+                if mismatch_ratio is None:
+                    mismatch_ratio = 1.0
+                mismatch = np.full(
+                    self.num_sites, mismatch_ratio * np.median(recombination)
+                )
+
+        if len(recombination) != max(0, self.num_sites - 1):
```

Comment:
Can use ``num_intervals`` here

Reply:
Great spot. Left over from when I didn't define that up top. Thanks.